src/Symfony/Component/HttpClient/Psr18Client.php

```diff
@@ -98,13 +98,26 @@ public function sendRequest(RequestInterface $request): ResponseInterface
             }
 
             $headers = $request->getHeaders();
-            if (!$request->hasHeader('content-length') && 0 <= $size = $body->getSize() ?? -1) {
+            $size = $request->getHeader('content-length')[0] ?? -1;
+            if (-1 === $size && 0 <= $size = $body->getSize() ?? -1) {
                 $headers['Content-Length'] = [$size];
             }
 
             $options = [
                 'headers' => $headers,
-                'body' => static fn (int $size) => $body->read($size),
+                'body' => 0 <= $size && $size < 1 << 21 ? $body->getContents() : static function (int $size) use ($body) {
```

Comment:
Can this logic also be applied to `HttplugClient`?

Reply:
sure, updated !